Library/Homebrew/utils/curl.rb

```diff
@@ -40,28 +61,50 @@ def curl_args(*extra_args, **options)
 
       args << "--show-error"
 
-      args << "--user-agent" << case options[:user_agent]
+      args << "--user-agent" << case user_agent
       when :browser, :fake
         HOMEBREW_USER_AGENT_FAKE_SAFARI
       when :default, nil
         HOMEBREW_USER_AGENT_CURL
       when String
-        options[:user_agent]
+        user_agent
+      else
+        raise TypeError, ":user_agent must be :browser, :fake, :default, or a String"
       end
 
       args << "--header" << "Accept-Language: en"
 
-      unless options[:show_output] == true
+      unless show_output == true
         args << "--fail"
         args << "--progress-bar" unless Context.current.verbose?
         args << "--verbose" if Homebrew::EnvConfig.curl_verbose?
         args << "--silent" unless $stdout.tty?
       end
 
-      args << "--connect-timeout" << connect_timeout.round(3) if options[:connect_timeout]
-      args << "--max-time" << max_time.round(3) if options[:max_time]
-      args << "--retry" << Homebrew::EnvConfig.curl_retries unless options[:retry] == false
-      args << "--retry-max-time" << retry_max_time.round if options[:retry_max_time]
+      if connect_timeout.is_a?(Numeric)
+        args << "--connect-timeout" << connect_timeout.round(3)
+      elsif connect_timeout.present?
+        raise TypeError, ":connect_timeout must be Numeric"
```

Comment:
I think given Sorbet and this being a private API it might be OK to just omit these type errors? Thoughts?

Reply:
By default, `sorbet-runtime` is only enabled in `brew tests` at the moment (via [`ENV["HOMEBREW_SORBET_RUNTIME"] = "1"`](https://github.com/Homebrew/brew/blob/b253b304d925bef2582c60eda22342e35a8205ba/Library/Homebrew/dev-cmd/tests.rb#L94)), so Sorbet only handles the type errors in that context. Outside of `brew tests`, we would rely on these `raise` statements and that's why I left them in.

After this is merged, I'll be creating a PR to allow for configuration of `curl` args/options in `livecheck` blocks. This is necessary for a variety of use cases (configuring the user agent, `POST` requests, etc.) and the values will be passed through livecheck to `#curl_args`.

With this in mind, I think it's better to raise an error, so the contributor will know that something isn't working as expected when they run `brew livecheck` on the formula/cask. Otherwise, `#curl_args` would quietly omit the argument on a non-numeric value and that can be a problem if it doesn't cause the check to noticeably fail. Even if the check does fail as a result, it wouldn't be immediately clear why it isn't working as expected and these errors are useful in that regard as well.

Reply:
> By default, `sorbet-runtime` is only enabled in `brew tests` at the moment (via [`ENV["HOMEBREW_SORBET_RUNTIME"] = "1"`](https://github.com/Homebrew/brew/blob/b253b304d925bef2582c60eda22342e35a8205ba/Library/Homebrew/dev-cmd/tests.rb#L94)), so Sorbet only handles the type errors in that context. Outside of `brew tests`, we would rely on these `raise` statements and that's why I left them in (after manually testing the behavior).

Sorbet will only handle runtime but it's still doing full-codebase static analysis. I think that's sufficient enough to remove the `TypeError`s (or can these values be set by non-maintainers?)

Reply:
My concern is about values provided indirectly to `#curl_args` from outside of the Homebrew/brew codebase (e.g., formulae/casks in first- and third-party taps). In the near future, we'll have `curl` args/options coming from a method in the `livecheck` block DSL but this can be addressed by doing the type checking in that method instead. We already do this with the other `livecheck` block DSL methods, so it makes sense to continue with that approach.

This would allow us to rely on Sorbet to do type checking for `#curl_args` but `Utils::Curl` is currently `typed: false` and there are ~40 issues to resolve before we can make it `true`. Until then, `brew typecheck` won't be enforcing argument types for `#curl_args` (and `sorbet-runtime` is only enabled for tests, so that won't help). [I can eventually work on making `Curl` typed as part of the other `Curl` work I have on the back burner but it won't be until I've finished with some forthcoming livecheck stuff (i.e., not soon).]

Removing the `raise` statements wouldn't be any worse than the status quo before this PR. If we do, we should also loosen the related conditions from `connect_timeout.is_a?(Numeric)` to `connect_timeout.present?`, so a bad value will cause a noticeable failure (in `#round` or from `curl`) instead of quietly omitting an argument.

This seems reasonable to me, so I'll update this accordingly.

Reply:
> My concern is about values provided indirectly to `#curl_args` from outside of the Homebrew/brew codebase (e.g., formulae/casks in first- and third-party taps).

I think it's worth having these checks in the public methods rather than this private one so the backtrace makes more sense.

> This would allow us to rely on Sorbet to do type checking for `#curl_args` but `Utils::Curl` is currently `typed: false` and there are ~40 issues to resolve before we can make it `true`. Until then, `brew typecheck` won't be enforcing argument types for `#curl_args` (and `sorbet-runtime` is only enabled for tests, so that won't help). [I can eventually work on making `Curl` typed as part of the other `Curl` work I have on the back burner but it won't be until I've finished with some forthcoming livecheck stuff (i.e., not soon).]

👍🏻 


> Removing the `raise` statements wouldn't be any worse than the status quo before this PR. If we do, we should also loosen the related conditions from `connect_timeout.is_a?(Numeric)` to `connect_timeout.present?`, so a bad value will cause a noticeable failure (in `#round` or from `curl`) instead of quietly omitting an argument.

I defer to you here @samford, merge whatever you feel is best (including this PR as-is, if desired).

Reply:
> I think it's worth having these checks in the public methods rather than this private one so the backtrace makes more sense.

Agreed. The errors from the `livecheck` block method are likely to be more understandable to contributors/maintainers than something coming from `#curl_args` 👍

> I defer to you here @samford, merge whatever you feel is best (including this PR as-is, if desired).

I removed the `raise` statements and loosened the conditions from `#is_a?` to `#present?` in the latest push. I imagine this should be fine, as some of these options didn't even function before and no one noticed until now 😅

Switching to named parameters and adding a method signature will automatically produce better documentation for `#curl_args`, so folks will be able to see the expected type for these parameters. Even with `Utils::Curl` being `typed: false`, this is an improvement over the existing situation and it will improve further when we can switch to `typed: true` in the future.